### ISSUE TITLE
SSP-Schema V1.0.0 and V1.0.1 Exceptions

### ISSUE THREAD

**Issue by mpemy** [usnistgov/OSCAL#1127]
# Describe the bug
I am building an OSCAL Converter to Word application using C# under .Net Framework 4.7.2 with Visual Studio 2019.
When I load SSP Schema to validate my oscal XML file I have an exception. This is the exception message:
 **System.Xml.Schema.XmlSchemaException:** 'The Pattern constraining facet is invalid - parsing "^([\p{_xmlI}-[:𐀀-󯿿]][\p{_xmlC}-[:𐀀-󯿿]]*)$" - [x-y] range in reverse order.'

This exception occurs when I try to compile the schema set.
These are the code lines where the exception occur
  XmlSchemaSet schema = new XmlSchemaSet();
                    schema.Add(XMLNamespace, XsdSchemaPath);
                    schema.Compile(); 

# Who is the bug affecting?
Tool vendors who are working to develop OSCAL-based tools and utilities.
# What is affected by this bug?
Because of this bug, I cannot validate if the oscal XML file my application creates is a valid v1.0.0 oscal file or a valid v1.0.1. But when I load the RC2 schema I do not have this exception.
{Describe the impact the bug is having.}
My application cannot Validate the OSCAL V1.0.0 files it produces against the official OSCAL Schema.
# When does this occur?
Once you load the schema and try to compile it, the IDE fire an exception
{Describe the conditions under which the bug is occurring.}

# How do we replicate the issue?
1. Start a basic console application with C# under .Net Framework 4.7.2
2. Load the ssp schema with the namespace  XMLNamespace = @"http://csrc.nist.gov/ns/oscal/1.0", once you compile the schema with the method Compile(), you get the exception. This is true for XML schema in the folder.  This is the file in question

```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using OSCALConversion;
using OSCALToolKitClasses;
using System.IO;
using System.Xml;
using System.Xml.Schema;
using System.Xml.Linq;

namespace OSCALConversionConsole
{
    class Program
    {
        static string sspSchema = @"C:\Users\Mous\Downloads\oscal-1.0.1\oscal-1.0.1\xml\schema\oscal_ssp_schema.xsd";
        static string sapSchema = @"C:\Users\Mous\Downloads\oscal-1.0.1\oscal-1.0.1\xml\schema\oscal_assessment-plan_schema.xsd";
        static string sarSchema = @"C:\Users\Mous\Downloads\oscal-1.0.1\oscal-1.0.1\xml\schema\oscal_assessment-results_schema.xsd";
        static string poamSchema = @"C:\Users\Mous\Downloads\oscal-1.0.1\oscal-1.0.1\xml\schema\oscal_poam_schema.xsd";
        static string completeSchema = @"C:\Users\Mous\Downloads\oscal-1.0.1\oscal-1.0.1\xml\schema\oscal_complete_schema.xsd";

        protected const string XMLNamespace = @"http://csrc.nist.gov/ns/oscal/1.0";
        static void Main(string[] args)
        {
            try
            {
                XmlSchemaSet schema = new XmlSchemaSet();
                schema.Add(XMLNamespace, completeSchema);
                schema.Compile();
            }
            catch (Exception exs)
            {
                string mess = exs.Message;
            }
        }
    }
}
```

If applicable, add screenshots to help explain your problem.}

# Expected behavior (i.e. solution)

The Pattern constraining facet is invalid - parsing "^([\p{_xmlI}-[:𐀀-󯿿]][\p{_xmlC}-[:𐀀-󯿿]]*)$" - [x-y] range in reverse order.

# Other Comments

{Add any other context about the problem here.}

**Comment by aj-stein-nist** [comment]
OK I confirmed I was able to reproduce the bug. More research is needed by me to determine the root cause and if/how we fix it. It does seem [this regex is the culprit](https://github.com/usnistgov/OSCAL/blob/870c4af2a7e8be41aaa729507e151f14b4de8924/xml/schema/oscal_ssp_schema.xsd#L3909) but not clear as to why it is acting up in C#. I will make a repro repo in GitHub eventually to make it easily testable and (potentially) work out a solution.

I will also keep quick, informal status updates in [the same Gitter thread following the 1.0.1 release](https://github.com/usnistgov/OSCAL/blob/870c4af2a7e8be41aaa729507e151f14b4de8924/xml/schema/oscal_ssp_schema.xsd#L3909).

**Comment by mpemy** [comment]
Thanks for the update.

**Comment by wendellpiez** [comment]
@mpemy @aj-stein-nist this is kind of interesting. Not our bug but we can work around it.

Short version: testing suggests that the error is the processor not the schema, but we could still relax the rule a little with no bad effect, writing an expression that works. Suggest we go back to `[\i-[:]][\c-[:]]*` or alternatively `[\w-[\d]]\w*`.

Longer version:

Given a regex `[\i-[:&#x10000;-&#xeffff;]][\c-[:&#x10000;-&#xeffff;]]*`, the error reported is what comes back if the regex is evaluated without the numeric character references (entity syntax) properly expanded (by the expected XML parse). In other words, if the literal `&#x10000`; is not expanded into the corresponding Unicode character (being excluded), and the processor sees "&#" instead of a single character starting the range, the expression can't be read.

Since this literal appears in an XSD file, which is XML, it really ought to be expanded before evaluation. (As happens in other XSD engines.) In fact the error message as reported by @mpemy is showing the characters, not the entity syntax (that's what the odd boxes and broken characters are in the paste). So it is being parsed for display, but not parsed for evaluation?

However, this could also be beside the point, if we have a way to avoid it, and testing today show that both the (unproblematic) expressions above also work to exclude names we don't want. So we run minimal or no risk to rewrite this.

I made an Issue in the Metaschema repo for this as linked above. (Thanks @mpemy! we can also help you patch this until schemas can be updated.)

**Comment by aj-stein-nist** [comment]
Hey, @wendellpiez, sorry for the delay this evening on a simple repro case. I made a quick separate repo. Is this what you had in mind?

https://github.com/aj-stein-nist/Issue1127Example

**Comment by aj-stein-nist** [comment]
@wendellpiez has been working on this after I made some reproduction code. I am going to work with Wendell to review the PR usnistgov/metaschema#183 and update this accordingly once merged in that repo _and_ the sub-module is updated in this repository and the changes propagate up.

@mpemy we have a fix in place _under review_, but you should be able to test the interim changes in your own software by patching in the recommended updates to the schema that we are proposing to fix these issues, separate of the review of the unit tests and other supporting work [in this snippet of the PR's commit for this change](https://github.com/usnistgov/metaschema/pull/183/files#diff-e39eba652060f8be0bfffa5c73d186e3fde53b651ccb1f80f7297c08fc091a26R164).

NOTE: As we discussed before, we built a quick test case library in C# exercising this updated XSD regex, and it already passes test with the relevant C# and `System.Xml` library in [a unit test](https://github.com/aj-stein-nist/Issue1127Example/blob/master/Regexted.Tests/SchemaPattern12.cs) and [basic XSD and XML documents](https://github.com/aj-stein-nist/Issue1127Example/commit/690924bd32fc7c359172db65105e489c07f24f8d#diff-ed7b8ac268901452b6a13a8c801abce353ff736a9f2fcbb0f50834412776c177R12) to test out this specific pattern, so whether you manually edit or wait for the merged code and bug release, I am fairly confident it will get you around the exception thrown in your case. Let us know!

**Comment by wendellpiez** [comment]
This Issue will be addressed by https://github.com/usnistgov/metaschema/pull/183 and can be closed when that PR is merged and its metaschema modifications are brought into OSCAL.

**Comment by david-waltermire-nist** [comment]
This work requires usnistgov/metaschema#191, which won't be ready until OSCAL 1.1. Moving the milestone accordingly.

**Comment by aj-stein-nist** [comment]
But isn't that a systematic rewrite and upgrade? With usnistgov/metaschema#183 and usnistgov/OSCAL#1161, this can be fixed for 1.0.2 for a fix release.

**Comment by aj-stein-nist** [comment]
@david-waltermire-nist, per issue triage this morning around the WIP PR and the touch-up work to cherry-pick usnistgov/metaschema#191 commits to move this up and close it separately, I will reassign the associated issue to you as well. That said, let me know if/when/how I can help.

1.0.2 or 1.0.1 release, I am not sure of the urgency. @mpemy has not got back to us on whether he was able to try to changes by switching submodules or modifying the code lines, so as discussed, I assume we can work on it at our pace.
